receiver/carbonreceiver/transport/tcp_server.go

```diff
@@ -0,0 +1,213 @@
+)
+
+const (
+	// TCPIdleTimeoutDefault is the default timeout for idle TCP connections.
+	TCPIdleTimeoutDefault = 30 * time.Second
+)
+
+type tcpServer struct {
+	ln          net.Listener
+	wg          sync.WaitGroup
+	idleTimeout time.Duration
+	reporter    Reporter
+}
+
+var _ (Server) = (*tcpServer)(nil)
+
+// NewTCPServer creates a transport.Server using TCP as its transport.
+func NewTCPServer(
+	addr string,
+	idleTimeout time.Duration,
+) (Server, error) {
+	if idleTimeout < 0 {
+		return nil, fmt.Errorf("invalid idle timeout: %v", idleTimeout)
+	}
+
+	if idleTimeout == 0 {
+		idleTimeout = TCPIdleTimeoutDefault
+	}
+
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	t := tcpServer{
+		ln:          ln,
+		idleTimeout: idleTimeout,
+	}
+	return &t, nil
+}
+
+func (t *tcpServer) ListenAndServe(
+	parser protocol.Parser,
+	nextConsumer consumer.MetricsConsumer,
+	reporter Reporter,
+) error {
+	if parser == nil || nextConsumer == nil || reporter == nil {
+		return errNilListenAndServeParameters
+	}
+
+	acceptedConnMap := make(map[net.Conn]struct{})
+	connMapMtx := &sync.Mutex{}
+
+	t.reporter = reporter
+	var err error
+	for {
+		conn, acceptErr := t.ln.Accept()
+		if acceptErr == nil {
+			connMapMtx.Lock()
+			acceptedConnMap[conn] = struct{}{}
+			connMapMtx.Unlock()
+			t.wg.Add(1)
+			go func(c net.Conn) {
+				t.handleConnection(parser, nextConsumer, c)
+				connMapMtx.Lock()
+				delete(acceptedConnMap, c)
+				connMapMtx.Unlock()
+				t.wg.Done()
+			}(conn)
+			continue
+		}
+
+		if netErr, ok := acceptErr.(net.Error); ok {
+			t.reporter.OnDebugf(
+				"TCP Transport (%s) - Accept (temporary=%v) net.Error: %v",
+				t.ln.Addr().String(),
+				netErr.Temporary(),
+				netErr)
+			if netErr.Temporary() {
+				continue
+			}
+		}
+
+		err = acceptErr
+		break
+	}
+
+	t.reporter.OnDebugf(
+		"TCP Transport (%s) exiting Accept loop error: %v",
+		t.ln.Addr().String(),
+		err)
+
+	// Close any lingering connection
+	connMapMtx.Lock()
+	for conn := range acceptedConnMap {
+		conn.Close()
+	}
+	connMapMtx.Unlock()
+
+	return err
+}
+
+func (t *tcpServer) Close() error {
+	err := t.ln.Close()
+	t.wg.Wait()
+	return err
+}
+
+func (t *tcpServer) handleConnection(
+	p protocol.Parser,
+	nextConsumer consumer.MetricsConsumer,
+	conn net.Conn,
+) {
+	defer conn.Close()
+	var span *trace.Span
+	reader := bufio.NewReader(conn)
+	for {
+		if span != nil {
+			span.End()
+			span = nil
+		}
+
+		if err := conn.SetDeadline(time.Now().Add(t.idleTimeout)); err != nil {
+			t.reporter.OnDebugf(
+				"TCP Transport (%s) - conn.SetDeadLine error: %v",
+				t.ln.Addr(),
+				err)
+			return
+		}
+		bytes, err := reader.ReadBytes((byte)('\n'))
+
+		// It is possible to have new data in bytes and err to be io.EOF
+		ctx, span := t.reporter.OnDataReceived(context.Background())
+		var numReceivedTimeSeries, numInvalidTimeSeries int
+		line := strings.TrimSpace(string(bytes))
```

Comment:
Is TrimSpace required? Can we have empty lines?

Reply:
Gateway precedent, ie.: implemented this way in Gateway - to be fair I didn't see it in any Carbon/Graphire spec.